testing/unit/test_copy.py

```diff
@@ -1,40 +1,158 @@
+        # Assert copy is a different object in memory
+        self.assertTrue(struc_copy is not struc)
+        self.validate_fields(struc, struc_copy, field_list)
+        # Validate no atoms or bonds copied
+        self.assertTrue(len(list(struc.atoms)) > 0)
+        self.assertTrue(len(list(struc_copy.atoms)) == 0)
+        self.assertTrue(len(list(struc.bonds)) > 0)
+        self.assertTrue(len(list(struc_copy.bonds)) == 0)
+
+    def test_shallow_copy_bonds(self):
+        struc = next(self.complex.bonds)
+        field_list = self.bond_fields
+        struc_copy = struc._shallow_copy()
+        self.assertTrue(struc_copy is not struc)
+        self.validate_fields(struc, struc_copy, field_list)
+
+    def test_shallow_copy_chain(self):
+        struc = next(self.complex.chains)
+        field_list = self.chain_fields
+        struc_copy = struc._shallow_copy()
+        self.assertTrue(struc_copy is not struc)
+        self.validate_fields(struc, struc_copy, field_list)
+        # Validate no residues copied
+        self.assertTrue(len(list(struc.residues)) > 0)
+        self.assertTrue(len(list(struc_copy.residues)) == 0)
+
+    def test_shallow_copy_molecule(self):
+        struc = next(self.complex.molecules)
+        field_list = self.molecule_fields
+        struc_copy = struc._shallow_copy()
+        self.assertTrue(struc_copy is not struc)
+        self.validate_fields(struc, struc_copy, field_list)
+        # Validate no chains copied
+        self.assertTrue(len(list(struc.chains)) > 0)
+        self.assertTrue(len(list(struc_copy.chains)) == 0)
+
+    def test_shallow_copy_complex(self):
+        struc = self.complex
+        field_list = self.complex_fields
+        struc_copy = struc._shallow_copy()
+        self.assertTrue(struc_copy is not struc)
+        self.validate_fields(struc, struc_copy, field_list)
+        # Validate no molecules copied
+        self.assertTrue(len(list(struc_copy.molecules)) == 0)
 
-    def test_shallow(self):
-        shallow_copy_tester(struct.Atom)
-        shallow_copy_tester(struct.Bond)
-        shallow_copy_tester(struct.Residue)
-        shallow_copy_tester(struct.Chain)
-        shallow_copy_tester(struct.Molecule)
-        shallow_copy_tester(struct.Complex)
-
-    def test_deep(self):
-        deep_copy_tester(2)
-        deep_copy_tester(3)
-        deep_copy_tester(4)
-        deep_copy_tester(5)
+    def test_deep_copy_complex(self):
+        comp = self.complex
+        field_list = self.complex_fields
+        comp_copy = comp._deep_copy()
+        # Traverse data structure and make sure all data was copied.
+        self.assertTrue(comp_copy is not comp)
+        self.validate_fields(comp, comp_copy, field_list)
+        for mol, mol_copy in zip(comp.molecules, comp_copy.molecules):
+            self.assertTrue(mol is not mol_copy)
+            self.validate_fields(mol, mol_copy, self.molecule_fields)
+            self.assertTrue(len(list(mol.chains)) > 0)
```

Comment:
instead of checking `len > 0` here and for chains, residues, and atoms, you should make sure the `len` is the same for the original and the copy. maybe also add a loop for bonds?